MultiSpeaker TTS with external variable size embedding

Edresson opened the pull request:
In this PR we implemented the paper
[Transfer Learning from Speaker Verification to Multispeaker Text-To-Speech Synthesis](https://papers.nips.cc/paper/7700-transfer-learning-from-speaker-verification-to-multispeaker-text-to-speech-synthesis.pdf). So this PR allows you to train a multi-speaker model that you can generalize to new speakers without needing extra training.

In addition, bug fixing was done in the GST implementation of the PR (#451 ) and unit tests for the GST were added.

My initial experiments were carried out in PR #394, but Mozilla TTS underwent several changes, so it was necessary to reimplement my best experiments.

lexkoro committed:
Added support for Tacotron2 GST + abbility to condition style input with wav or tokens

lexkoro committed:
No need to query every token when none were passed

lexkoro committed:
fix fft_size key error

lexkoro committed:
pylint

lexkoro committed:
tacotrongst test + test fixes

lexkoro committed:
Merge remote-tracking branch 'upstream/dev' into dev

lexkoro committed:
override compute_gst in tacotron2 model

lexkoro committed:
small gst config change

erogol committed:
update comment

Edresson committed:
merge dev into branch

Edresson committed:
add support fot VCTK and BRSpeech dataset

Edresson committed:
add Portuguese Cleaner

Edresson committed:
bugfix in DDC now DDC work on Tacotron1

Edresson committed:
add External Embedding per sample instead of nn.Embedding

Edresson committed:
Merge branch 'dev' into dev-gst-embeddings

Edresson committed:
bugfix in GST

Edresson committed:
fix Lint check

Edresson committed:
travis unit tests fix and add Tacotron and Tacotron 2 GST and MultiSpeaker Tests

Edresson committed:
add support for synthesize using variable size external embedding and add bugfix in scipy.io import

Edresson commented:
Below are Colabs notebooks for Demo of two models, the demos allow upload of GST samples, upload of wav file to use as a reference for the speaker (synthesize with your own voice).

[Demo Mozilla TTS MultiSpeaker Jia et al. 2018](https://colab.research.google.com/drive/1OX2jtyxmeFt7kRO65QgPy65ku6HWUrFK?usp=sharing)
[Demo Mozilla TTS MultiSpeaker Jia et al. 2018 With GST](https://colab.research.google.com/drive/1JX29gDcsPHx5HlbwNkpHnPQkhhrSvrji?usp=sharing)

Both models were trained using the VCTK dataset. All checkpoints, configs and logs are available on the Google Drive:
[Mozilla TTS MultiSpeaker Jia et al. 2018](https://drive.google.com/drive/folders/15R0XOs5iTckNIaP7pKpYaTFwxW4rT1CD?usp=sharing)
[Mozilla TTS MultiSpeaker Jia et al. 201 With GST](https://drive.google.com/drive/folders/15HE-5dfPi7AwVta0FfUimNOwaOLifNVX?usp=sharing)

The models synthesize well, but when changing the sample of extraction of the embeddings the synthesized voice is not very close to the reference voice. Perhaps the problem is the Speaker Encoder. I think we should train with more data and try to use the Angular Prototypical that achieved state of the art results in  the paper ["In defence of metric learning for speaker recognition"](https://arxiv.org/abs/2003.11982)

Edresson commented:
I implemented the Angular Prototycal loss function and trained a SpeakerEncoder with Angular Prototycal loss. The Model was trained using the Libri-TTS dataset (using the train-clean-360 and train-clean-100 subsets), and was trained by 77.982 steps. For the training I used the Speaker-Encoder-iter25k model (transfered with GE2E loss) as transfer learning and just continued the training by changing the loss function to AngleProto and changing the extraction parameters of the spectrograms to be compatible with the TTS Multi-Speaker model(the two checkpoints I have made available here).
The logs and the best checkpoint of the model are available on [Google Drive](https://drive.google.com/drive/folders/1I_aZ0z0j0EEyZviQixzxegWiz19LDQIN?usp=sharing) 

My idea was to use this model and calculate an extra loss function between the embedding of the expected spectrogram and the spectrogram predicted by the network. At the moment I don't have to do it in maybe a few weeks!

Edresson committed:
Force the loading of the wav in synthesis using the AP sample rate to avoid breaking the demo

Edresson committed:
add gst suport in synthesize

Edresson committed:
add gst style dict  suport in synthesize

Edresson committed:
bugfix in tacotron and tacotron 2 gst inference

Edresson committed:
added integrity test for GST dictionary in synthesis

Edresson committed:
add Colab Notebook from TTS Multi-Speaker

Edresson committed:
clean Colab Notebook outputs

Edresson committed:
add Colab Notebook from TTS Multi-Speaker with GST

Edresson committed:
add Jupyter Notebook for Extract Speaker Embedding per sample using GE2E

Edresson committed:
Ops! Map Notebooks imports for TTS current version

Edresson committed:
Implement Angular Prototypical loss

Edresson committed:
add suport for AngleProto loss

Edresson committed:
update speaker encoder config.json, for compatibility with the TTS model

Edresson committed:
fix broken imports in speaker encoder

Edresson committed:
update deprecated functions  call from speaker encoder

Edresson committed:
add speaker encoder parameters in config.json, for more easy changes

Edresson committed:
add Jupyter Notebook for Extract Speaker Embedding per sample using AngleProto

Edresson committed:
fix Lint check errors

Edresson committed:
add test for AngleProtoLoss

Edresson committed:
fix Lint check errors

Edresson committed:
Merge branch 'dev' into dev-gst-embeddings

Edresson committed:
fix Lint check after merge

erogol commented:
Would you mind If I put your example to https://github.com/mozilla/TTS/wiki/TTS-Notebooks-and-Tutorials ?

erogol commented:
If you like you can also provide a shell script for training this model and we can place it under https://github.com/erogol/TTS_recipes/blob/master/README.md

You can also use your fork for this if you like.

erogol commented:
great results on your notebook samples. One suggestion is that you can stop asking synthesis if the user provides 'q' as input like 'quit'.

erogol commented on TTS/bin/train_tts.py:
```diff
@@ -485,22 +496,40 @@ def main(args):  # pylint: disable=redefined-outer-name
-                                                   "introduce new speakers to " \
-                                                   "a previously trained model."
-        else:
+            if c.use_external_speaker_embedding_file: # if restore checkpoint and use External Embedding file
```
Can we place this speaker embedding loading part into a function so that we can re-use it for a different model or training run?

Edresson replied:
I'm a little confused. I don't see another place to use this, because this code snippet uses load_speaker_mapping and just checks if there is a file speakers.json in args.restore_path, if it exists it will be loaded, otherwise it tries to load c.external_speaker_embedding_file.
Do you see any other use for this, even though we already have load_speaker_mapping?
If so what part do you want me to put in a function?

erogol replied:
for instance, if we use a different model and write a different train_new_model.py instead of copying and pasting this part it would be better to import this part as a function? Does it make sense?

Edresson replied:
I understand now ! Makes sense :)

erogol commented on TTS/tts/datasets/TTSDataset.py:
```diff
@@ -70,7 +72,7 @@ def __init__(self,
 
     def load_wav(self, filename):
-        audio = self.ap.load_wav(filename)
+        audio = self.ap.load_wav(filename, sr=self.sample_rate)
```
Let's do not resample audio here. It slows down the data loading a lot. User should resample audio in advance for better performance.

Edresson replied:
OK!  I fixed this

Edresson replied:
> @Edresson Nice results so far!
> 
> One question though, any findings on your side how robust the scaling of tokens or providing a ref. wav is regarding the attention? With ddc also new masking was introduced [L130](https://github.com/Edresson/TTS/blob/e20fac2ed34e7f21e54b8d0f1668f2d3a50b30ec/TTS/tts/models/tacotron2.py#L130) and [L136](https://github.com/Edresson/TTS/blob/e20fac2ed34e7f21e54b8d0f1668f2d3a50b30ec/TTS/tts/models/tacotron2.py#L136).
> In my previous tests those maskings would mostly break the attention when scaling a token.
> I've since retrained without the masking and the model seems more robust + flexible when conditioning on style tokens.
> 
> At last I have to say that using ddc in combination with gst has produces so far the best quality + robustness of model I've trained so far.

 My results with GST, do not allow many modifications. As Embeddings speakers are not good enough if I keep a lot of GST tokens, GST starts to learn the characteristics of the speaker. This way, GST did not allow generalizing to new speakers the only way I found was to drastically reduce the number of GST tokens (I used only 5 tokens). So my GST control results are not good. I believe that if we train a speaker encoder with much more data (Maybe using LibriTTS, VoxCeleb1 and VoxCeleb2), we can have a more accurate representation of each speaker and train a multispeaker model with a more robust GST.

In my tests scaling of tokens the model is more robust, when we use a .wav reference, depending on the sample the speaker speaks very fast, or even does not speak the first word!

Thanks for the tips :). It's good to know what worked best for you

erogol commented on TTS/tts/layers/gst_layers.py:
```diff
@@ -96,7 +96,7 @@ def __init__(self, num_heads, num_style_tokens,
         self.style_tokens = nn.Parameter(
             torch.FloatTensor(num_style_tokens, self.key_dim))
-        nn.init.orthogonal_(self.style_tokens)
+        nn.init.normal_(self.style_tokens, mean=0, std=0.5)
```
Why do you use normal initialization? Orthogonal initialization should work better since we expect each style token to be different than the other.

Edresson replied:
@erogol  I used normal initialization, because [Mellotron](https://github.com/NVIDIA/mellotron/blob/master/modules.py) uses it, as it was the repository with the most expressive results using GST, I found it interesting to keep what they implemented. If you still wanted, I can make the change!

erogol replied:
not really. If it is the case let it stay that way. Thx for the info :)

Edresson replied:
Ok :)

erogol commented on TTS/tts/models/tacotron.py:
```diff
@@ -37,37 +40,52 @@ def __init__(self,
-        proj_speaker_dim = 80 if num_speakers > 1 else 0
-        # base model layers
+                             ddc_r, gst, gst_embedding_dim, gst_num_heads, gst_style_tokens)
+
```
Is there any multi-speaker part can be places under tacotron_abstract.py?

Edresson replied:
I made the changes

SanjaESC commented:
@Edresson Nice results so far! 

One question though, any findings on your side how robust the scaling of tokens or providing a ref. wav is regarding the attention?  With ddc also new masking was introduced [L130](https://github.com/Edresson/TTS/blob/e20fac2ed34e7f21e54b8d0f1668f2d3a50b30ec/TTS/tts/models/tacotron2.py#L130) and [L136](https://github.com/Edresson/TTS/blob/e20fac2ed34e7f21e54b8d0f1668f2d3a50b30ec/TTS/tts/models/tacotron2.py#L136). 
In my previous tests those maskings would mostly break the attention when scaling a token. 
I've since retrained without the masking and the model seems more robust + flexible when conditioning on style tokens. 

At last I have to say that using ddc in combination with gst has produces so far the best quality + robustness of model I've trained so far.

erogol commented on TTS/tts/models/tacotron2.py:
```diff
@@ -28,35 +27,60 @@ def __init__(self,
-        proj_speaker_dim = 80 if num_speakers > 1 else 0
-        # base layers
+                             ddc_r, gst, gst_embedding_dim, gst_num_heads, gst_style_tokens)
+
```
The same here maybe can share some stuff in the abstract class.

Edresson replied:
I made the changes

Edresson replied:
I didn't remember that!
I just sent the commit that implements these tests!

erogol commented on tests/test_layers.py:
```diff
@@ -71,38 +70,6 @@ def test_in_out():
         assert output.shape[2] == 2, "size not {}".format(output.shape[2])
         assert stop_tokens.shape[0] == 4
 
```
Can you implement some tests for multi-speaker model? At least running it with a dummy input.

Edresson replied:
I didn't remember that!
I just sent the commit that implements these tests!

Edresson replied:
let me know if I need to do anything else!

erogol reviewed (changes requested):
I add some feedback. I hope these are not demotivating after introducing this big PR.

Edresson committed:
do not resample audio in Dataloader

Edresson commented:
> @Edresson Nice results so far!
> 
> One question though, any findings on your side how robust the scaling of tokens or providing a ref. wav is regarding the attention? With ddc also new masking was introduced [L130](https://github.com/Edresson/TTS/blob/e20fac2ed34e7f21e54b8d0f1668f2d3a50b30ec/TTS/tts/models/tacotron2.py#L130) and [L136](https://github.com/Edresson/TTS/blob/e20fac2ed34e7f21e54b8d0f1668f2d3a50b30ec/TTS/tts/models/tacotron2.py#L136).
> In my previous tests those maskings would mostly break the attention when scaling a token.
> I've since retrained without the masking and the model seems more robust + flexible when conditioning on style tokens.
> 
> At last I have to say that using ddc in combination with gst has produces so far the best quality + robustness of model I've trained so far.

My results with GST, do not allow many modifications. As Embeddings speakers are not good enough if I keep a lot of GST tokens, GST starts to learn the characteristics of the speaker. This way, GST did not allow generalizing to new speakers the only way I found was to drastically reduce the number of GST tokens (I used only 5 tokens). So my GST control results are not good. I believe that if we train a speaker encoder with much more data (Maybe using LibriTTS, VoxCeleb1 and VoxCeleb2), we can have a more accurate representation of each speaker and train a multispeaker model with a more robust GST.

In my tests scaling of tokens the model is more robust, when we use a .wav reference, depending on the sample the speaker speaks very fast, or even does not speak the first word!

Thanks for the tips :). It's good to know what worked best for you

erogol commented:
@Edresson what prevents you to train a better speaker encoder? GPUs?

Edresson commented:
> what prevents you to train a better speaker encoder? GPUs?

Yes, I currently have no GPUs for training. In addition we have to improve the speaker encoder dataloader, as it is currently  slow.

I also didn’t have time to try to support the checkpoint provided in the Repository [Real-Time-Voice-Cloning](https://github.com/CorentinJ/Real-Time-Voice-Cloning) the TTS speaker encoder, this would bring a great Speaker Encoder and allow us to achieve much better results.

Edresson committed:
use tacotron abstract for multispeaker common definitions

Edresson committed:
add tests for tacotron and tacotron2 Multi-speaker

erogol commented:
> 
> 
> > what prevents you to train a better speaker encoder? GPUs?
> 
> Yes, I currently have no GPUs for training. In addition we have to improve the speaker encoder dataloader, as it is currently slow.
> 
> I also didn’t have time to try to support the checkpoint provided in the Repository [Real-Time-Voice-Cloning](https://github.com/CorentinJ/Real-Time-Voice-Cloning) in the TTS speaker encoder, this would bring a great Speaker Encoder and allow us to achieve much better results.

I can provide you a hand for GPUs if you can provide a single shell script that would train the model.

Edresson commented:
> > > what prevents you to train a better speaker encoder? GPUs?
> > 
> > 
> > Yes, I currently have no GPUs for training. In addition we have to improve the speaker encoder dataloader, as it is currently slow.
> > I also didn’t have time to try to support the checkpoint provided in the Repository [Real-Time-Voice-Cloning](https://github.com/CorentinJ/Real-Time-Voice-Cloning) in the TTS speaker encoder, this would bring a great Speaker Encoder and allow us to achieve much better results.
> 
> I can provide you a hand for GPUs if you can provide a single shell script that would train the model.

Do you want a script that trains the Speaker Encoder or the TTS MultiSpeaker?

Edresson committed:
add support for CorentinJ Speaker encoder and add notebook for extract embeddings

Edresson commented:
I implemented CorentinJ Speaker Encoder Support and converted the checkpoint so we can use that checkpoint for finetunning in future training. The[ CorentinJ model](https://github.com/CorentinJ/Real-Time-Voice-Cloning) was trained by 1564501 steps in the VoxCeleb1, VoxCeleb 2 and LibriSpeech datasets.

I believe that the use of this Speaker Encoder can bring great improvements to the multispeaker model!

GE2E CorentinJ Converted Checkpoint : https://drive.google.com/drive/folders/1U2EM_vkE0CD4ShEwfT80XMdM9wPvnakl?usp=sharing

erogol commented:
> 
> 
> I implemented CorentinJ Speaker Encoder Support and converted the checkpoint so we can use that checkpoint for finetunning in future training. The[ CorentinJ model](https://github.com/CorentinJ/Real-Time-Voice-Cloning) was trained by 1564501 steps in the VoxCeleb1, VoxCeleb 2 and LibriSpeech datasets.
> 
> I believe that the use of this Speaker Encoder can bring great improvements to the multispeaker model!
> 
> GE2E CorentinJ Converted Checkpoint : https://drive.google.com/drive/folders/1U2EM_vkE0CD4ShEwfT80XMdM9wPvnakl?usp=sharing

do you have any example to back this ?

Edresson commented:
> > I implemented CorentinJ Speaker Encoder Support and converted the checkpoint so we can use that checkpoint for finetunning in future training. The[ CorentinJ model](https://github.com/CorentinJ/Real-Time-Voice-Cloning) was trained by 1564501 steps in the VoxCeleb1, VoxCeleb 2 and LibriSpeech datasets.
> > I believe that the use of this Speaker Encoder can bring great improvements to the multispeaker model!
> > GE2E CorentinJ Converted Checkpoint : https://drive.google.com/drive/folders/1U2EM_vkE0CD4ShEwfT80XMdM9wPvnakl?usp=sharing
> 
> do you have any example to back this ?

I'm training a model with this speaker encoder. I extract the embedding using this [notebook. ](https://github.com/Edresson/TTS/blob/dev-gst-embeddings/notebooks/GE2E-CorentinJ-ExtractSpeakerEmbeddings-by-sample.ipynb)
I converted the checkpoint with the intention of using it in the future to initialize the model weights, since the extraction of Spectrograms between this Speaker Encoder and the TTS are not compatible, therefore we cannot add it at the moment.

erogol commented:
I merged this to dev branch with all the renaming and refactoring stuff. So closing this PR. Thanks again for this :).

Please continue from the latest branch if you like to continue for more.

erogol commented:
Due to a lot of refactoring in dev, I rebased your PR. (I am not sure I did good). 

Let's continue by #489 

I close this one. Hope it is alright

kunalch204 commented:
> Below are Colabs notebooks for Demo of two models, the demos allow upload of GST samples, upload of wav file to use as a reference for the speaker (synthesize with your own voice).
> 
> [Demo Mozilla TTS MultiSpeaker Jia et al. 2018](https://colab.research.google.com/drive/1OX2jtyxmeFt7kRO65QgPy65ku6HWUrFK?usp=sharing) [Demo Mozilla TTS MultiSpeaker Jia et al. 2018 With GST](https://colab.research.google.com/drive/1JX29gDcsPHx5HlbwNkpHnPQkhhrSvrji?usp=sharing)
> 
> Both models were trained using the VCTK dataset. All checkpoints, configs and logs are available on the Google Drive: [Mozilla TTS MultiSpeaker Jia et al. 2018](https://drive.google.com/drive/folders/15R0XOs5iTckNIaP7pKpYaTFwxW4rT1CD?usp=sharing) [Mozilla TTS MultiSpeaker Jia et al. 201 With GST](https://drive.google.com/drive/folders/15HE-5dfPi7AwVta0FfUimNOwaOLifNVX?usp=sharing)
> 
> The models synthesize well, but when changing the sample of extraction of the embeddings the synthesized voice is not very close to the reference voice. Perhaps the problem is the Speaker Encoder. I think we should train with more data and try to use the Angular Prototypical that achieved state of the art results in the paper ["In defence of metric learning for speaker recognition"](https://arxiv.org/abs/2003.11982)

These links dont work anymore, any idea on where to access these?
